src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/costmodel/CFRNumberOfTermsShippedInResponses.java

```diff
@@ -61,26 +58,19 @@ else if ( lop instanceof LogicalOpBGPAdd ) {
 			}
 		}
 		else if ( lop instanceof LogicalOpRequest ) {
-			final DataRetrievalRequest req = ((LogicalOpRequest<?, ?>) lop).getRequest();
-
-			// TODO: Check whether the the following is correct. I believe that
-			// this part should be defined in terms of the type of federation
-			// members and it should be similar to the case of LogicalOpTPAdd.
+			final CompletableFuture<Integer> futureIntResSize = initiateCardinalityEstimation(plan);
+			final FederationMember fm = ((LogicalOpRequest<?, ?>) lop).getFederationMember();
 
-			if ( req instanceof TriplePatternRequest ) {
-				final TriplePatternRequest tpReq = (TriplePatternRequest) req;
-				final int numberOfVars = tpReq.getQueryPattern().numberOfVars();
-				return CompletableFuture.completedFuture( 3 * numberOfVars );
-			}
-			else if ( req instanceof SPARQLRequest ) {
-				final SPARQLRequest sparqlReq = (SPARQLRequest) req;
-				// TODO: The following line is not correct. Not every SPARQLRequest contains a BGP.
-				final int numberOfVars = QueryPatternUtils.getVariablesInPattern( (BGP) sparqlReq.getQueryPattern() ).size();
-				final CompletableFuture<Integer> futureIntResSize = initiateCardinalityEstimation(plan);
+			if ( fm instanceof SPARQLEndpoint ) {
+				final SPARQLRequest req = (SPARQLRequest) ((LogicalOpRequest<?, ?>) lop).getRequest();
+				final int numberOfVars = QueryPatternUtils.getVariablesInPattern( req.getQueryPattern() ).size();
 				return futureIntResSize.thenApply( intResSize -> numberOfVars * intResSize );
 			}
+			else if ( fm instanceof TPFServer || fm instanceof BRTPFServer) {
+				return futureIntResSize.thenApply( intResSize -> 3 * intResSize );
+			}
 			else {
-				throw createIllegalArgumentException(req);
+				throw createIllegalArgumentException(fm);
```

Comment:
Before throwing this exception, `futureIntResSize` should be canceled. See the case of `LogicalOpTPAdd` a bit further up in this source file.